cmd/thanos/store.go

```diff
@@ -186,6 +187,9 @@ func (sc *storeConfig) registerFlag(cmd extkingpin.FlagClause) {
 	cmd.Flag("store.enable-lazy-expanded-postings", "If true, Store Gateway will estimate postings size and try to lazily expand postings if it downloads less data than expanding all postings.").
 		Default("false").BoolVar(&sc.lazyExpandedPostingsEnabled)
 
+	cmd.Flag("store.streaming-series-limit", "The maximum series allowed to match when streaming series. The Series/LabelNames/LabelValues call fails if this limit is exceeded. 0 means no limit.").
```

Comment:
I wonder if this should be a limiter we apply when lazy postings are enabled, instead of through a separate flag. To me it seems like the store-gw should be able to choose the limiter itself and not rely on a config value.

Reply:
Yeah I think that makes sense to me. We can apply the limit in a streaming way when lazy postings or vertical sharding is enabled.

The only concern for me is that it changes the existing behavior. If a user don't want streaming series limit at all but they are using vertical sharding, then they don't have a way to restore previous behavior.

Reply:
Yeah vertical sharding is hard to solve because it's distributed. But query shards could also hit different store-gw instances, so I am not sure if a global series limit is meaningful.

Reply:
Ah I didn't mean to talk about global series limiter. Just saying in a single request to a single instanct, CX might still want to use the existing series limiter rather than the `streaming` one because existing series limiter covers how many series touched.

Reply:
I think for these cases, maybe we should make the limiter configurable in the constructor? In Thanos we can use the streaming limiter for lazy postings, and the original one for eager postings. And Cortex can decide to configure store-gw as it sees fit.

Reply:
That would be quite complex tbh. I want to keep it simple, too.
I am thinking about two options:
1. As you said we can just do streaming limiter for lazy postings and all other cases use original one. Ysers doesn't need to configure anything. Store Gateway will check whether it is lazy posting or not
2. Change everything to just use streaming limiter. Honestly I think this is fine as it is what we use for other Stores like sidecar, receiver, etc.

I think I will probably go with option 1 in this pr. WDYT?

Reply:
Yes, this sounds good to me 👍 Let's start with this solution and we can still change it in the future.

Reply:
Fixed. PTAL